### ISSUE TITLE
RLMException: 'The Realm is already in a write transaction' when deleting records

### ISSUE THREAD

**Issue by torfluor** [mentrena/SyncKit#148]
Sometimes when I delete an object in my app I get the following exception: RLMException', reason: 'The Realm is already in a write transaction'

I looks like the issue is related to the following: 
In the RealmSwiftAdapter.swift deleteRecords method line 945  beginWrite() is called. Before this write ends a new beginWrite() is called on line 370 in the method updateTracking in the same file.

My guess is that there is a notification that is triggering the updateTracking method when the object is deleted. 

The following GitHub tread might be useful: https://github.com/realm/realm-cocoa/issues/4511

Maybe checking realm.isInWriteTransaction in updateTracking will solve this issue?

Here is the stack trace:

```
#0	0x00000001900f79d8 in __exceptionPreprocess ()
#1	0x00000001a447ab54 in objc_exception_throw ()
#2	0x000000010092ce8c in RLMSetErrorOrThrow(NSError*, NSError* __autoreleasing*) at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/realm-cocoa/Realm/RLMUtil.mm:380
#3	0x00000001008bace4 in RLMRealmTranslateException(NSError* __autoreleasing*) at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/realm-cocoa/Realm/RLMRealm.mm:347
#4	0x00000001008bda38 in -[RLMRealm beginWriteTransactionWithError:] at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/realm-cocoa/Realm/RLMRealm.mm:612
#5	0x00000001008bd9d4 in -[RLMRealm beginWriteTransaction] at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/realm-cocoa/Realm/RLMRealm.mm:603
#6	0x00000001009a426c in Realm.beginWrite() at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/realm-cocoa/RealmSwift/Realm.swift:285
#7	0x00000001006bce5c in RealmSwiftAdapter.updateTracking(objectIdentifier:entityName:inserted:deleted:changes:realmProvider:) at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/SyncKit/SyncKit/Classes/RealmSwift/RealmSwiftAdapter.swift:370
#8	0x00000001006bc988 in closure #1 in RealmSwiftAdapter.updateTracking(insertedObject:identifier:entityName:provider:) at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/SyncKit/SyncKit/Classes/RealmSwift/RealmSwiftAdapter.swift:349
#9	0x00000001006bca5c in partial apply for closure #1 in RealmSwiftAdapter.updateTracking(insertedObject:identifier:entityName:provider:) ()
#10	0x00000001009304a8 in closure #1 in RLMObjectBase._observe<A>(on:_:) at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/realm-cocoa/RealmSwift/Aliases.swift:111
#11	0x00000001009308c4 in thunk for @escaping @callee_guaranteed (@guaranteed RLMObjectBase?, @guaranteed [String]?, @guaranteed [Any]?, @guaranteed [Any]?, @guaranteed Error?) -> () ()
#12	0x00000001007a8400 in (anonymous namespace)::ObjectChangeCallbackWrapper::after(realm::CollectionChangeSet const&) at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/realm-cocoa/Realm/RLMObjectBase.mm:588
#13	0x00000001007a735c in realm::CollectionChangeCallback::Impl2<(anonymous namespace)::ObjectChangeCallbackWrapper>::after(realm::CollectionChangeSet const&) at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/realm-core/src/realm/object-store/collection_notifications.hpp:214
#14	0x0000000100a43880 in realm::CollectionChangeCallback::after(realm::CollectionChangeSet const&) at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/realm-core/src/realm/object-store/collection_notifications.hpp:143
#15	0x0000000100a437f4 in auto realm::_impl::CollectionNotifier::after_advance()::$_9::operator()<realm::util::CheckedUniqueLock, realm::_impl::CollectionNotifier::Callback>(realm::util::CheckedUniqueLock&, realm::_impl::CollectionNotifier::Callback&) const at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/realm-core/src/realm/object-store/impl/collection_notifier.cpp:349
#16	0x0000000100a235cc in void realm::_impl::CollectionNotifier::for_each_callback<realm::_impl::CollectionNotifier::after_advance()::$_9>(realm::_impl::CollectionNotifier::after_advance()::$_9&&) at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/realm-core/src/realm/object-store/impl/collection_notifier.cpp:397
#17	0x0000000100a234d4 in realm::_impl::CollectionNotifier::after_advance() at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/realm-core/src/realm/object-store/impl/collection_notifier.cpp:338
#18	0x0000000100a247fc in realm::_impl::NotifierPackage::after_advance() at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/realm-core/src/realm/object-store/impl/collection_notifier.cpp:498
#19	0x0000000100a9e0d8 in void (anonymous namespace)::advance_with_notifications<realm::_impl::transaction::begin(std::__1::shared_ptr<realm::Transaction> const&, realm::BindingContext*, realm::_impl::NotifierPackage&)::$_2>(realm::BindingContext*, std::__1::shared_ptr<realm::Transaction> const&, realm::_impl::transaction::begin(std::__1::shared_ptr<realm::Transaction> const&, realm::BindingContext*, realm::_impl::NotifierPackage&)::$_2&&, realm::_impl::NotifierPackage&) at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/realm-core/src/realm/object-store/impl/transact_log_handler.cpp:601
#20	0x0000000100a9dd28 in realm::_impl::transaction::begin(std::__1::shared_ptr<realm::Transaction> const&, realm::BindingContext*, realm::_impl::NotifierPackage&) at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/realm-core/src/realm/object-store/impl/transact_log_handler.cpp:636
#21	0x0000000100a559d0 in realm::_impl::RealmCoordinator::promote_to_write(realm::Realm&) at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/realm-core/src/realm/object-store/impl/realm_coordinator.cpp:1160
#22	0x0000000100ba1c04 in realm::Realm::begin_transaction() at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/realm-core/src/realm/object-store/shared_realm.cpp:637
#23	0x00000001008bda08 in -[RLMRealm beginWriteTransactionWithError:] at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/realm-cocoa/Realm/RLMRealm.mm:608
#24	0x00000001008bd9d4 in -[RLMRealm beginWriteTransaction] at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/realm-cocoa/Realm/RLMRealm.mm:603
#25	0x00000001009a426c in Realm.beginWrite() at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/realm-cocoa/RealmSwift/Realm.swift:285
#26	0x00000001006c9008 in closure #1 in RealmSwiftAdapter.deleteRecords(with:) at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/SyncKit/SyncKit/Classes/RealmSwift/RealmSwiftAdapter.swift:945
#27	0x00000001006b389c in closure #1 in executeOnMainQueue(_:) at /Users/tor/Library/Developer/Xcode/DerivedData/Timeline_Creator_Pro-gavjgorfgjhnbzcqksdfhwghdhzf/SourcePackages/checkouts/SyncKit/SyncKit/Classes/RealmSwift/RealmSwiftAdapter.swift:17
#28	0x00000001006a6fd0 in thunk for @callee_guaranteed () -> () ()
#29	0x00000001006a7028 in thunk for @escaping @callee_guaranteed () -> () ()
#30	0x00000001049d9528 in _dispatch_client_callout ()
#31	0x00000001049e88a8 in _dispatch_async_and_wait_invoke ()
#32	0x00000001049d9528 in _dispatch_client_callout ()
#33	0x00000001049e7994 in _dispatch_main_queue_callback_4CF ()
#34	0x00000001900775e0 in __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ ()
#35	0x0000000190071a88 in __CFRunLoopRun ()
#36	0x0000000190070ba0 in CFRunLoopRunSpecific ()
#37	0x00000001a6dd6598 in GSEventRunModal ()
#38	0x00000001929622f4 in -[UIApplication _run] ()
#39	0x0000000192967874 in UIApplicationMain ()
#40	0x00000001004bee20 in main at /Users/tor/dev/timeline-creator-pro/Timeline Creator Pro/Misc/AppDelegate.swift:19
#41	0x000000018fd4f568 in start ()
```

**Comment by torfluor** [comment]
Today I also got this issue when updating an object. Then it crashes as link 346 in RealmSwiftAdapter.swift in the updateTracking method.

**Comment by mentrena** [comment]
Hey! That thread was useful, can you have a look at https://github.com/mentrena/SyncKit/pull/149 ?
It changes `updateTracking` to make it safe for this scenario. Let me know if it works.

**Comment by torfluor** [comment]
Hi. I tried to compile with the fixes-1.0.1 branch, but I get errors that CloudKitSynchronizer has no member cloudSharingControllerDidChangeShare and cloudSharingControllerDidDeleteShare. Has this been removed?

**Comment by torfluor** [comment]
Hi. I have tested with the fixes-1.0.1 branch, and I have not seen the exception. Thank you for fixing this :)
